src/Microsoft.Health.Dicom.Core/Features/Update/UpdateInstanceService.cs

```diff
@@ -0,0 +1,202 @@
+        _metadataStore = EnsureArg.IsNotNull(metadataStore, nameof(metadataStore));
+        _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+        _recyclableMemoryStreamManager = EnsureArg.IsNotNull(recyclableMemoryStreamManager, nameof(recyclableMemoryStreamManager));
+    }
+
+    /// <inheritdoc />
+    public async Task UpdateInstanceBlobAsync(InstanceFileIdentifier instanceFileIdentifier, DicomDataset datasetToUpdate, CancellationToken cancellationToken)
+    {
+        EnsureArg.IsNotNull(datasetToUpdate, nameof(datasetToUpdate));
+        EnsureArg.IsNotNull(instanceFileIdentifier, nameof(instanceFileIdentifier));
+        EnsureArg.IsTrue(instanceFileIdentifier.NewVersion.HasValue, nameof(instanceFileIdentifier.NewVersion.HasValue));
+
+        Task updateInstanceFileTask = UpdateInstanceFileAsync(instanceFileIdentifier, datasetToUpdate, cancellationToken);
+        Task updateInstanceMetadataTask = UpdateInstanceMetadataAsync(instanceFileIdentifier, datasetToUpdate, cancellationToken);
+        await Task.WhenAll(updateInstanceFileTask, updateInstanceMetadataTask);
+    }
+
+    /// <inheritdoc />
+    public async Task DeleteInstanceBlobAsync(long fileIdentifier, CancellationToken cancellationToken = default)
+    {
+        _logger.LogInformation("Begin deleting instance blob {FileIdentifier}.", fileIdentifier);
+
+        Task fileTask = _fileStore.DeleteFileIfExistsAsync(fileIdentifier, cancellationToken);
+        Task metadataTask = _metadataStore.DeleteInstanceMetadataIfExistsAsync(fileIdentifier, cancellationToken);
+
+        await Task.WhenAll(fileTask, metadataTask);
+
+        _logger.LogInformation("Deleting instance blob {FileIdentifier} completed successfully.", fileIdentifier);
+    }
+
+    private async Task UpdateInstanceFileAsync(InstanceFileIdentifier instanceFileIdentifier, DicomDataset datasetToUpdate, CancellationToken cancellationToken)
+    {
+        long originFileIdentifier = instanceFileIdentifier.Version;
+        long newFileIdentifier = instanceFileIdentifier.NewVersion.Value;
+
+        KeyValuePair<string, long> block = default;
+
+        // If the file is already updated, then we can copy the file and just update the first block
+        if (instanceFileIdentifier.OriginalVersion.HasValue)
+        {
+            _logger.LogInformation("Begin copying instance file {OrignalFileIdentifier} - {NewFileIdentifier}", originFileIdentifier, newFileIdentifier);
+            await _fileStore.CopyFileAsync(originFileIdentifier, newFileIdentifier, cancellationToken);
+        }
+        else
+        {
+            _logger.LogInformation("Begin downloading original file {OrignalFileIdentifier} - {NewFileIdentifier}", originFileIdentifier, newFileIdentifier);
+
+            using Stream stream = await _fileStore.GetFileAsync(originFileIdentifier, cancellationToken);
+            DicomFile dcmFile = await DicomFile.OpenAsync(stream, FileReadOption.ReadLargeOnDemand);
```

Comment:
can we skip reading largeObjects?

Reply:
We need to read large objects to create a copy of the file